ddtrace/internal/telemetry/telemetry_writer.py

```diff
@@ -0,0 +1,213 @@
+        """
+        Sends a telemetry request to telemetry intake [this will be switched the agent]
+        """
+        conn = get_connection(self.url)
+        rb_json = self.encoder.encode(request["body"])
+        resp = None
+        with StopWatch() as sw:
+            try:
+                conn.request("POST", self.url, rb_json, request["headers"])
+                resp = get_connection_response(conn)
+
+                t = sw.elapsed()
+                if t >= self.interval:
+                    log_level = logging.WARNING
+                else:
+                    log_level = logging.DEBUG
+
+                log.log(log_level, "sent %d in %.5fs to %s. response: %s", len(rb_json), t, self.url, resp.status)
+            finally:
+                conn.close()
+        return resp
+
+    def flush_integrations_queue(self):
+        # type () -> List[Dict]
+        """Returns a list of all integrations queued by integration_event"""
+        with self._lock:
+            integrations = self._integrations_queue
+            self._integrations_queue = []
+        return integrations
+
+    def flush_events_queue(self):
+        # type () -> List[TelemetryRequest]
+        """Returns a list of all integrations queued by classmethods"""
+        with self._lock:
+            requests = self._events_queue
+            self._events_queue = []
+        return requests
+
+    def queued_events(self):
+        # type () -> List[TelemetryRequest]
+        return self._events_queue
+
+    def queued_integrations(self):
+        # type () -> List[Dict]
+        return self._integrations_queue
+
+    def periodic(self):
+        integrations = self.flush_integrations_queue()
+        if integrations:
+            integrations_request = app_integrations_changed_telemetry_request(integrations)
+            TelemetryWriter.add_event(integrations_request)
+
+        requests = self.flush_events_queue()
+        requests_failed = []  # type: List[TelemetryRequest]
+        for request in requests:
+            resp = self._send_request(request)
+            if resp.status >= 300:
+                requests_failed.append(request)
+
+        with self._lock:
+            # add failed requests back to the requests queue
+            self._events_queue.extend(requests_failed)
+
+    def shutdown(self):
+        # type: () -> None
+        self.app_closed_event()
+        self.periodic()
+
+    @classmethod
+    def add_event(cls, request):
+        # type: (TelemetryRequest) -> None
+        """
+        Adds a Telemetry Request to the TelemetryWriter request buffer
+
+         :param TelemetryRequest request: dictionary which stores a formatted telemetry request body and header
+        """
+        with cls._lock:
+            if cls._instance is None:
+                return
+            request["body"]["seq_id"] = cls.sequence
+            cls.sequence += 1
+            # TO DO: replace list with a dead letter queue
```

Comment:
why?

Reply:
Previously there was some weirdness where I'd try to set the seq_id in multiple places (ex. app-started event always had a sequence id of 0). This field should only be set be set when a TelemetryRequest is added to the events queue.